File: p151_175/p156/digitcounter.py
import time

class DigitCounter:
  def __init__(self, limit):
    self.limit = limit
    self.counts = [0] * 10
    self.sum_of_solutions = 0

  def get_sum(self):
    solution_counter = 1
    for i in xrange(0, (self.limit + 1)):
      for ch in list(str(i)):
        digit = (int)(ch)
        self.counts[digit] += 1

      is_solution = False
      for digit in xrange(1, 10):
        if self.counts[digit] == i:
          print solution_counter, i, digit
          self.sum_of_solutions += i
          is_solution = True

      if is_solution:
        solution_counter += 1

    return self.sum_of_solutions

def main():
  start_time = time.time()
  limit = 10 ** 6
  digit_counter = DigitCounter(limit)
  print digit_counter.get_sum()
  end_time = time.time()
  print end_time - start_time

if __name__ == '__main__':
  main()

File: p151_175/p156/src/DigitCounter.java
public class DigitCounter {
	private long limit;
	private long[] counts = new long[10];
	private long sumOfSolutions = 0;
	
	public DigitCounter(long startLimit){
		limit = startLimit;
	}
	
	public long getSum(){
		int solutionCounter = 1;
		for (long i = 0; i <= limit; i++) {
			if (i % 1000000000L == 0) {
				System.out.println(i);
			}
			String iString = String.valueOf(i); 
			for (int charIndex=0; charIndex < iString.length(); charIndex++) {
				char c = iString.charAt(charIndex);
				int digitIndex = Character.getNumericValue(c);
				if (digitIndex > 9 || digitIndex < 0) {
					System.out.println(i);
					System.out.println(c);
				}
				counts[digitIndex]++;
			}
			
			boolean isSolution = false;
			for(int digitIndex = 1; digitIndex <= 9; digitIndex++){
				if (counts[digitIndex] == i) {
					System.out.printf("%d %d digit: %d\n", solutionCounter, i, digitIndex);
					sumOfSolutions += i;
					isSolution = true;
				}
			}
			if (isSolution) {
				solutionCounter++;
			}
		}
		
		return sumOfSolutions;
	}
}
